module/x/gravity/keeper/attestation_handler.go

```diff
@@ -21,317 +22,510 @@ var _ types.DistributionKeeper = (*distrkeeper.Keeper)(nil)
+	coin := sdk.NewCoin(denom, claim.Amount)
+	coins := sdk.Coins{coin}
+
+	moduleAddr := a.keeper.accountKeeper.GetModuleAddress(types.ModuleName)
+	if !isCosmosOriginated { // We need to mint eth-originated coins (aka vouchers)
+		preMintBalance := a.keeper.bankKeeper.GetBalance(ctx, moduleAddr, denom)
+		// Ensure that users are not bridging an impossible amount, only 2^256 - 1 tokens can exist on ethereum
+		prevSupply := a.keeper.bankKeeper.GetSupply(ctx, denom)
+		newSupply := new(big.Int).Add(prevSupply.Amount.BigInt(), claim.Amount.BigInt())
+		if newSupply.BitLen() > 256 { // new supply overflows uint256
+			a.keeper.logger(ctx).Error("Deposit Overflow",
 				"claim type", claim.GetType(),
-				"id", types.GetAttestationKey(claim.GetEventNonce(), hash),
 				"nonce", fmt.Sprint(claim.GetEventNonce()),
 			)
-			return sdkerrors.Wrap(errTokenAddress, "invalid token contract on claim")
+			return sdkerrors.Wrap(types.ErrIntOverflowAttestation, "invalid supply after SendToCosmos attestation")
 		}
-		if errEthereumSender != nil {
+
+		if err := a.keeper.bankKeeper.MintCoins(ctx, types.ModuleName, coins); err != nil {
+			// in this case we have lost tokens! They are in the bridge, but not
+			// in the community pool our out in some users balance, every instance of this
+			// error needs to be detected and resolved
 			hash, _ := claim.ClaimHash()
-			a.keeper.logger(ctx).Error("Invalid ethereum sender",
-				"cause", errEthereumSender.Error(),
+			a.keeper.logger(ctx).Error("Failed minting",
+				"cause", err.Error(),
 				"claim type", claim.GetType(),
 				"id", types.GetAttestationKey(claim.GetEventNonce(), hash),
 				"nonce", fmt.Sprint(claim.GetEventNonce()),
 			)
-			return sdkerrors.Wrap(errTokenAddress, "invalid ethereum sender on claim")
+			return sdkerrors.Wrapf(err, "mint vouchers coins: %s", coins)
 		}
 
-		// While not strictly necessary, explicitly making the receiver a native address
-		// insulates us from the implicit address conversion done in x/bank's account store iterator
-		nativeReceiver, err := types.GetNativePrefixedAccAddress(receiverAddress)
+		postMintBalance := a.keeper.bankKeeper.GetBalance(ctx, moduleAddr, denom)
+		if postMintBalance.Sub(preMintBalance).Amount != claim.Amount {
+			panic(fmt.Sprintf(
+				"Somehow minted incorrect amount! Previous balance %v Post-mint balance %v claim amount %v",
+				preMintBalance.String(), postMintBalance.String(), claim.Amount.String()),
+			)
+		}
+	}
 
+	if !invalidAddress { // address appears valid, attempt to send minted/locked coins to receiver
+		preSendBalance := a.keeper.bankKeeper.GetBalance(ctx, moduleAddr, denom)
+		err := a.sendCoinToCosmosAccount(ctx, claim, receiverAddress, coin)
 		if err != nil {
 			invalidAddress = true
+		} else {
+			postSendBalance := a.keeper.bankKeeper.GetBalance(ctx, moduleAddr, denom)
+			if preSendBalance.Sub(postSendBalance).Amount != claim.Amount {
```

Comment:
we have a reversible context here (if my reading of `processAttestation` is correct, can we return an error rather than panicking? I guess it depends on our confidence that this current execution wil be rolled back in that case.